File: server/agent/reasoning/alpha_beta.py
import math
from server.agent.knowledge.evaluation import evaluate


# =========================================
# ALPHA-BETA CORE
# =========================================

def alphabeta(board, depth, alpha, beta, maximizing):

    # Base case
    status = board.game_status()
    if depth == 0 or status != "ongoing":
        return evaluate(board)

    moves = board.get_all_legal_moves()

    if maximizing:

        max_eval = -math.inf

        for move in moves:

            board.make_move(move)

            eval = alphabeta(board, depth - 1, alpha, beta, False)

            board.undo_move()

            max_eval = max(max_eval, eval)
            alpha = max(alpha, eval)

            # PRUNE
            if beta <= alpha:
                break

        return max_eval

    else:

        min_eval = math.inf

        for move in moves:

            board.make_move(move)

            eval = alphabeta(board, depth - 1, alpha, beta, True)

            board.undo_move()

            min_eval = min(min_eval, eval)
            beta = min(beta, eval)

            # PRUNE
            if beta <= alpha:
                break

        return min_eval


# =========================================
# BEST MOVE FINDER
# =========================================

def get_best_move(board, depth):

    best_move = None

    maximizing = True if board.turn == "white" else False

    alpha = -math.inf
    beta = math.inf

    moves = board.get_all_legal_moves()

    # OPTIONAL: simple move ordering (captures first)
    moves.sort(key=lambda m: m.piece_captured != "", reverse=True)

    if maximizing:

        max_eval = -math.inf

        for move in moves:

            board.make_move(move)

            eval = alphabeta(board, depth - 1, alpha, beta, False)

            board.undo_move()

            if eval > max_eval:
                max_eval = eval
                best_move = move

            alpha = max(alpha, eval)

    else:

        min_eval = math.inf

        for move in moves:

            board.make_move(move)

            eval = alphabeta(board, depth - 1, alpha, beta, True)

            board.undo_move()

            if eval < min_eval:
                min_eval = eval
                best_move = move

            beta = min(beta, eval)

    return best_move

File: server-java/src/main/java/com/pace/chess/engine/AlphaBeta.java
package com.pace.chess.engine;

import com.pace.chess.model.Board;
import com.pace.chess.model.Move;
import com.pace.chess.model.Piece;
import java.util.List;

public class AlphaBeta {

    public static Move getBestMove(Board board, int depth) {
        // 'mathColor' flips the evaluation perspective (1 for White, -1 for Black)
        int mathColor = (board.turn == Piece.WHITE) ? 1 : -1;
        
        double alpha = Double.NEGATIVE_INFINITY;
        double beta = Double.POSITIVE_INFINITY;

        // Generate and filter legal moves directly
        List<Move> pseudoMoves = MoveGenerator.generateAllMoves(board, board.turn);
        List<Move> moves = RuleChecker.filterLegalMoves(board, pseudoMoves, board.turn);
        
        MoveSorter.sortMoves(moves, board);

        Move bestMove = null;
        double bestValue = Double.NEGATIVE_INFINITY;

        for (Move m : moves) {
            board.makeMove(m);
            // Evaluation is always from White's perspective, negamax flips it
            double value = -negamax(board, depth - 1, -beta, -alpha, -mathColor);
            board.undoMove();

            if (value > bestValue) {
                bestValue = value;
                bestMove = m;
            }
            alpha = Math.max(alpha, value);
        }
        return bestMove;
    }

    private static double negamax(Board board, int depth, double alpha, double beta, int mathColor) {
        List<Move> pseudoMoves = MoveGenerator.generateAllMoves(board, board.turn);
        List<Move> moves = RuleChecker.filterLegalMoves(board, pseudoMoves, board.turn);

        // PERFORMANCE FIX: Check if game is over without expensive separate calls
        if (moves.isEmpty()) {
            if (RuleChecker.isInCheck(board, board.turn)) {
                return -1000000; // Checkmate: Current player loses
            }
            return 0; // Stalemate: Draw
        }

        // Base case: return the static evaluation
        if (depth == 0) return mathColor * Evaluation.evaluate(board);

        MoveSorter.sortMoves(moves, board);

        double max = Double.NEGATIVE_INFINITY;
        for (Move m : moves) {
            board.makeMove(m);
            double score = -negamax(board, depth - 1, -beta, -alpha, -mathColor);
            board.undoMove();

            max = Math.max(max, score);
            alpha = Math.max(alpha, score);
            if (alpha >= beta) break; // Alpha-Beta Pruning
        }
        return max;
    }
}